lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemStats.java

```diff
@@ -0,0 +1,263 @@
+                Optional<String> operationName = metricCollection.metricValues(OPERATION_NAME).stream().filter(Objects::nonNull).findFirst();
+                if (serviceId.isEmpty() || operationName.isEmpty()) {
+                    log.warn("ServiceId or OperationName is empty for AWS MetricCollection: %s", metricCollection);
+                    return;
+                }
+
+                if (!serviceId.get().equals("S3")) {
+                    return;
+                }
+
+                AwsSdkV2ApiCallStats apiCallStats = getApiCallStats(operationName.get());
+                publishMetrics(metricCollection, apiCallStats);
+            }
+            catch (Exception e) {
+                log.warn(e, "Publishing AWS metrics failed");
+            }
+        }
+
+        private void publishMetrics(MetricCollection metricCollection, AwsSdkV2ApiCallStats apiCallStats)
+        {
+            metricCollection.stream()
+                    .filter(metricRecord -> metricRecord.value() != null && ALLOWED_METRICS.contains(metricRecord.metric()))
+                    .forEach(metricRecord -> {
+                        if (metricRecord.metric().equals(API_CALL_SUCCESSFUL)) {
+                            Boolean value = (Boolean) metricRecord.value();
+
+                            stats.total.updateCalls();
+                            apiCallStats.updateCalls();
+
+                            if (value.equals(Boolean.FALSE)) {
+                                stats.total.updateFailures();
+                                apiCallStats.updateFailures();
+                            }
+                        }
+                        else if (metricRecord.metric().equals(RETRY_COUNT)) {
+                            int value = (int) metricRecord.value();
+
+                            stats.total.updateRetries(value);
+                            apiCallStats.updateRetries(value);
+                        }
+                        else if (metricRecord.metric().equals(API_CALL_DURATION)) {
+                            Duration value = (Duration) metricRecord.value();
+
+                            stats.total.updateLatency(value);
+                            apiCallStats.updateLatency(value);
+                        }
+                        else if (metricRecord.metric().equals(ERROR_TYPE)) {
+                            String value = (String) metricRecord.value();
+
+                            if (value.equals(THROTTLING.toString())) {
+                                stats.total.updateThrottlingExceptions();
+                                apiCallStats.updateThrottlingExceptions();
+                            }
+                            else if (value.equals(SERVER_ERROR.toString())) {
+                                stats.total.updateServerErrors();
+                                apiCallStats.updateServerErrors();
+                            }
```

Comment:
can there be other errors which we should account for?

Reply:
Yes there are errors other than server error (non-throttling) and throttling in [here](https://github.com/aws/aws-sdk-java-v2/blob/master/core/sdk-core/src/main/java/software/amazon/awssdk/core/metrics/CoreMetric.java#L171) but I thought these 2 are the major errors and we can add other errors incrementally if needed.

Reply:
I would say we probably don't want to loose error stats, so we should add `otherErrors` and/or could all categories from https://github.com/aws/aws-sdk-java-v2/blob/master/core/sdk-core/src/main/java/software/amazon/awssdk/core/metrics/CoreMetric.java#L171